### PR TITLE
fix(output): gate content-block recognition on each field's value shape

### PR DESCRIPTION
**Pull request by alexander-turner**
Claims `CONTENT_BLOCK_SCHEMAS` / `isContentBlock` in `src/output.mjs`. Follow-up to #401 (merged as 2.57.2), which landed the content-block collapse for #398. This carries the hardening that review of #401 produced but that did not make it into the merge — the adversarial pass, plus the two findings its reviewers left on the merged commit.

## What & why

Recognise a content block by each field's **value shape**, not its key name alone; stop recognising the four tags that pair a block with another block; and name the guard that keeps a polluted prototype from faking a tag.

**Key names alone are not a block.** `source`, `title` and `content` are generic field names, so `{ type: "image", source: "https://cdn/x.png" }` — an ordinary record of a URL — matched the image schema and was collapsed, losing its fields. A real image block's `source` is an object. Each schema field now carries a predicate on its value (`text` a string, `source` an object, `citations` an array for a text block but the `{ enabled }` object for a document), and a mismatch falls back to the leaf-wise walk. This is the repo's own rule: gate on the value's shape, not the name.

**Collapsing a paired block re-creates the bug it fixes.** `tool_use`, `server_tool_use`, `tool_result` and `web_search_tool_result` carry an id binding them to a partner block. Replacing one with a text block orphans that partner, and the API rejects the request exactly as permanently as the rewritten `type` tag #398 is about. They are dropped from the schema map and keep the walk, so **#398 still stands for those four tags** — a negative test says so in the open rather than trading one instance of the bug for another. (This is also the answer to Codex's P1 on #401 about `web_fetch_tool_result` / `code_execution_tool_result`: both are `tool_use_id`-paired, so they are excluded by the same rule, not by oversight.)

**The own-key `type` check was an unnamed, unpinned security guard.** It reads as redundant beside the schema-map miss, so a cleanup would delete it — and nothing went red when it did: every negative case decided on the map lookup instead. With `Object.prototype.type` polluted, `value.type` resolves through the prototype, so `{ text: "leak" }` would tag itself a text block and be collapsed, dropping a legitimate field on the fail-closed path. The invariant is now stated at the enforcement point and a test kills the mutant.

## Partitions

Three commits, one concern, read in order:

1. `e02029c` — value-shape predicates; drop the pairing-bearing tags.
2. `040df29` — name and pin the prototype-pollution guard; trim the collapse comment to the 5-line cap.

## Review focus

`src/output.mjs` — `CONTENT_BLOCK_SCHEMAS` and `isContentBlock`. Two things worth checking against the API docs rather than the diff:

- **The predicates themselves.** Too strict is fail-safe (that tag falls back to the walk); too loose collapses an object that isn't a block. The one I'd check first is `document.citations`, which is the `{ enabled }` toggle object while `text.citations` is an array — the two tags genuinely differ.
- **`isValidField` looks up predicates with `Object.hasOwn`**, mirroring `mapFilterWarning`'s guard in the same file: a bare index would resolve `toString`/`constructor` to real functions and let a key outside the schema pass as if it had a predicate.

## How it was tested

- `node --test test/output.test.mjs` → 174/174. The negative table covers unknown tag, missing required key, extra key, **wrong-shaped required value**, **wrong-shaped optional value**, non-string tag, and the two pairing-bearing tags — each asserting the walk still runs. Separate cases pin the `Array.isArray` guard (an array carrying own `type`/`text` properties stays an array) and the prototype guard.
- Non-vacuity, both new guards: dropping `!Object.hasOwn(value, "type")` reds exactly the pollution test (`# fail 1`); short-circuiting `isContentBlock` to `false` reds the collapse suite. Green with them.
- Depth-boundary cases either side of `MAX_DEPTH` pin the collapse's position before the depth guard: a block reached at exactly `MAX_DEPTH` collapses; one level deeper the subtree truncates to the bare sentinel.
- `pnpm check`, `pnpm lint`, `pnpm format:check` clean. `build-hook-binaries.mjs --check` reproduces the manifest byte-for-byte after the rebase; `plugin-hooks.bundle.mjs` conflicted on the rebase and was regenerated with the repo's tooling, never resolved by hand.
- `test/fuzz-coverage.test.mjs`'s exemption justification for `suppressToolOutput` moves in the same commit as the recogniser it describes, per the SSOT-contract rule.

## Decisions made

- **`tool_result` and friends keep the walk rather than preserving their pairing keys.** The alternative — keep `id`/`tool_use_id` and suppress only the payload — puts attacker-chosen strings back into a fail-closed placeholder. Leaving #398 unfixed for four rare tags is the cheaper failure.
- **Recognition stays context-free rather than positional.** Codex's other P1 on #401 asked for envelope/position context instead of matching every walked object. The value-shape gate answers the concrete misclassification it named (`{ type: "image", source: "…" }`); true positional context would mean threading a "this is a content position" bit through the walk, which the walk cannot know from the inside — `tool_response` is the tool's own schema, not a message body.
- **The schema map still covers the Messages API shapes only.** MCP tool results use their own block shapes (`{ type: "image", data, mimeType }`, `resource`, `resource_link`); nothing in this tree records which shape `tool_response` carries for an MCP result, so adding them would be unverified recall traded against the precision rule.

### CODE REVIEW

File: src/output.mjs [hunk 1]
```diff
@@ -1113,77 +1116,109 @@ export function suppressToolOutput(value, message) {
...
-    "web_search_tool_result",
-    { required: ["tool_use_id", "content"], optional: ["cache_control"] },
+    "thinking",
+    { required: { thinking: isString, signature: isString }, optional: {} },
   ],
-]);
+  ["redacted_thinking", { required: { data: isString }, optional: {} }],
+];
+
+const CONTENT_BLOCK_SCHEMAS = new Map(CONTENT_BLOCK_SCHEMA_ENTRIES);
```
[the remark this answers is not in the document]
Comment (alexander-turner) [reply src/output.mjs:1216]: The inline doesn't typecheck, so I took the intent a different way. Annotating the `new Map(...)` directly is what I wrote first; `tsc` rejects it:

```
Type '{ required: { text: FieldShape; source?: undefined; ... }; ... }' is not assignable to type 'BlockSchema'.
  Types of property 'required' are incompatible.
    Property 'source' is incompatible with index signature.
      Type 'undefined' is not assignable to type 'FieldShape'.
```

TypeScript unions the six entry literals before checking them against `Map<string, BlockSchema>`, and the union gives every entry the other entries' keys as `?: undefined`, which fails `Record<string, FieldShape>`'s index signature. The element-wise `[string, BlockSchema][]` annotation contextually types each literal on its own, so it's the only form that compiles.

What I did instead, since the underlying complaint — the documentation sits under an identifier nobody arrives at — is fair: the header now says why the separate const exists (so the next reader doesn't retry the inline), and `CONTENT_BLOCK_SCHEMAS` carries its own one-liner pointing at `isContentBlock`'s lookup.

---
_Generated by [Claude Code](https://claude.ai/code)_
